### ISSUE TITLE
Integration test only works if the app is already running in the background

### ISSUE THREAD

**Issue by davidxmoody** [AppsFlyerSDK/react-native-appsflyer#22]
Hi,

We are encountering some unusual behaviour and I'm hoping you can help. The problem we are having is that when using the [integration testing app](https://support.appsflyer.com/hc/en-us/articles/207032126-AppsFlyer-SDK-Integration-Android), it always fails on the first attempt. However, it works if my app is already running in the background. I'm not sure why this is happening and don't know if it will work in production or not (we have not released it yet).

Here are the things we did to set it up:

- Add `compile project(path: ':react-native-appsflyer')` to `android/app/build.gradle`
- Add the `SingleInstallBroadcastReceiver` to the `AndroidManifest.xml`
- Add the `@integer/google_play_services_version` meta attribute to the `AndroidManifest.xml`
- Add the import and `RNAppsFlyerPackage` package to `MainApplication.java`
- Include and set project dir in `android/settings.gradle`
- Call `initSdk` with our config (also call `getAppsFlyerUID`)
- Whitelist my device (it appears in the list)

Description of the issue:

- Running the "SDK Integration Test" from the provided integration testing app from the Google Play Store fails on the first attempt
    - Our AppsFlyer dashboard shows a click but no install
    - The integration testing app says there is a problem (the generic "This app can not be monitored." message)
- If you clear the app in the background (open app switcher, swipe it away) then the problem will repeat itself
- However, if you don't clear it and re-run the integration test again then it succeeds!
    - It successfully registers as both a click and non-organic install in the dashboard
    - Re-running it will continue to work until the app is cleared in the background
    - It also succeeds if the first launch was from manually opening the app rather than the integration testing app opening it

Other information that might be of use:

- I'm using a Nexus 6P
- I also tested it on an older HTC One and get the exact same behaviour
- `onInstallConversionData` appears to never call its callback in either the successful or unsuccessful case
- The logs say that neither the Android ID nor IMEI are being tracked
    - I thought that was the issue for a while and tried to manually set the Android ID but that doesn't fix anything
- I tried running them demo app and everything shows as "not initialized yet"/"not called yet"
    - There appear to be no relevant logs either
    - The integration testing app also does not work with it (and also complains about the missing intent stuff)

Do you know what's going on? I am very happy to provide more details or logs or try anything you can suggest.

Thanks,
David

**Comment by davidxmoody** [comment]
I'm still attempting to debug the problem but not getting very far. Here are some logs from the integration test in case they may be helpful. They are filtered by the string "I/AppsFlyer" using the Android Monitor in Android Studio (not sure if there is a better way or if I'm missing some other useful logs).

The first set are from the first launch using the integration testing app. I then waited a bit and re-ran it (without clearing our app from memory) which resulted in the second set of results. As before, the integration testing app said the first one failed and second one succeeded and the dashboard shows two clicks and one non-organic install.

```
# From first unsuccessful run (filtered by "I/AppsFlyer")

04-04 10:24:47.731 23183-23183/com.momentum.moneyhub.app I/AppsFlyer_4.7.1: ****** onReceive called *******
04-04 10:24:47.754 23183-23183/com.momentum.moneyhub.app I/AppsFlyer_4.7.1: Play store referrer: af_tranid=JoBX8Mff2nogaBVk6kvpxg&pid=AppsFlyer_Test&c=Integration_Test
04-04 10:24:47.820 23183-23183/com.momentum.moneyhub.app I/AppsFlyer_4.7.1: Test mode started..
04-04 10:24:48.804 22384-23099/? I/AppsFlyer_4.6.7: onBecameBackground
04-04 10:24:48.804 22384-23099/? I/AppsFlyer_4.6.7: callStatsBackground background call
04-04 10:24:48.805 22384-23099/? I/AppsFlyer_4.6.7: app went to background
04-04 10:24:48.808 22384-23099/? I/AppsFlyer_4.6.7: call = https://stats.appsflyer.com/stats; size = 430 bytes; body = {"advertiserId":"c03e4a93-2cfe-4959-9e0f-f2cc22eb4a3f","platform":"Android","gcd_conversion_data_timing":"0","devkey":"ud****************WMsN","statType":"user_closed_app","originalAppsflyerId":"1490949938374-6358470765804798556","time_in_app":"12","deviceFingerPrintId":"ffffffff-9333-3215-0000-00000625e8bb","app_id":"com.appsflyer.referrerSender","uid":"1490949938374-6358470765804798556","channel":null,"launch_counter":"249"}
04-04 10:24:48.844 22384-23099/? I/AppsFlyer_4.6.7: Status 200 ok
04-04 10:24:48.844 22384-23099/? I/AppsFlyer_4.6.7: Connection call succeeded
04-04 10:24:54.259 22384-22384/? I/AppsFlyer_4.6.7: onBecameForeground
04-04 10:24:54.304 22384-23308/? I/AppsFlyer_4.6.7: sendTrackingWithEvent from activity: com.appsflyer.referrerSender.ReferrerSender
04-04 10:24:54.305 22384-23308/? I/AppsFlyer_4.6.7: Trying to fetch GAID..
04-04 10:24:54.367 22384-23308/? I/AppsFlyer_4.6.7: ******* sendTrackingWithEvent: com.momentum.moneyhub.app
04-04 10:24:54.408 22384-23308/? I/AppsFlyer_4.6.7: IMEI was not collected.
04-04 10:24:54.410 22384-23309/? I/AppsFlyer_4.6.7: sendTrackingWithEvent from activity: com.appsflyer.referrerSender.ResultActivity
04-04 10:24:54.410 22384-23309/? I/AppsFlyer_4.6.7: Trying to fetch GAID..
04-04 10:24:54.411 22384-23308/? I/AppsFlyer_4.6.7: Android ID was not collected.
04-04 10:24:54.419 22384-23308/? I/AppsFlyer_4.6.7: AppsFlyer: first launch date: 2017-03-31_095752+0100
04-04 10:24:54.419 22384-23308/? I/AppsFlyer_4.6.7: AppsFlyerLib.sendTrackingWithEvent
04-04 10:24:54.422 22384-23308/? I/AppsFlyer_4.6.7: url: https://events.appsflyer.com/api/v4/androidevent?buildnumber=4.6.7&app_id=com.appsflyer.referrerSender
04-04 10:24:54.422 22384-23308/? I/AppsFlyer_4.6.7: data: {"device":"angler","firstLaunchDate":"2017-03-31_095752+0100","installDate":"2017-03-31_094538+0100","sdk":"25","referrer":"utm_source=google-play&utm_medium=organic","carrier":"3","deviceFingerPrintId":"ffffffff-9333-3215-0000-00000625e8bb","date1":"2017-03-31_094538+0100","af_preinstalled":"false","advertiserIdEnabled":"false","iaecounter":"83","appsflyerKey":"ud****************WMsN","lang_code":"en","installer_package":"com.android.vending","app_version_name":"4.2.0","lang":"English","timepassedsincelastlaunch":"18","eventName":"com.momentum.moneyhub.app","advertiserId":"c03e4a93-2cfe-4959-9e0f-f2cc22eb4a3f","isGaidWithGps":"true","deviceData":{"build_display_id":"N4F26T","arch":"","cpu_abi":"arm64-v8a","cpu_abi2":""},"af_v2":"0ad129bfd25e32926a8b8228f3accd0c97e8b377","deviceType":"user","af_v":"a44cc5b7aa9c3637a11fbb115bc3420d0aca68f4","app_version_code":"22","eventValue":"{\"Error Message\":\"This app can not be monitored. Please make sure that:\\n\\n1. You are using a supported AppsFlyer SDK version.\\n2. AppsFlyer's BroadcastReceiver is configured.\\n3. startTracking() is placed in the MainActivity\",\"Test-App Version\":\"4.2.0\"}","af_events_api":"1","platformextension":"android_native","network":"WIFI","operator":"","country":"GB","date2":"2017-03-31_094538+0100","brand":"google","prev_event":"{\"prev_event_timestamp\":\"1491297451907\",\"prev_event_value\":\"{\\\"Error Message\\\":\\\"This app can not be monitored. Please make sure that:\\\\n\\\\n1. You are using a supported AppsFlyer SDK version.\\\\n2. AppsFlyer's BroadcastReceiver is configured.\\\\n3. startTracking() is placed in the MainActivity\\\",\\\"Test-App Version\\\":\\\"4.2.0\\\"}\",\"prev_event_name\":\"com.momentum.moneyhub.app\"}","af_timestamp":"1491297894366","uid":"1490949938374-6358470765804798556","isFirstCall":"false","fb":"001217d2-982f-401d-9d3c-3ed8c9cc0eff","af_preinstall_name":"af","counter":"249","model":"Nexus 6P","product":"angler"}
04-04 10:24:54.427 22384-23309/? I/AppsFlyer_4.6.7: ******* sendTrackingWithEvent: Launch
04-04 10:24:54.461 22384-23309/? I/AppsFlyer_4.6.7: IMEI was not collected.
04-04 10:24:54.463 22384-23309/? I/AppsFlyer_4.6.7: Android ID was not collected.
04-04 10:24:54.468 22384-23309/? I/AppsFlyer_4.6.7: AppsFlyer: first launch date: 2017-03-31_095752+0100
04-04 10:24:54.468 22384-23309/? I/AppsFlyer_4.6.7: AppsFlyerLib.sendTrackingWithEvent
04-04 10:24:54.471 22384-23309/? I/AppsFlyer_4.6.7: url: https://t.appsflyer.com/api/v4/androidevent?buildnumber=4.6.7&app_id=com.appsflyer.referrerSender
04-04 10:24:54.471 22384-23309/? I/AppsFlyer_4.6.7: data: {"device":"angler","firstLaunchDate":"2017-03-31_095752+0100","installDate":"2017-03-31_094538+0100","sdk":"25","referrer":"utm_source=google-play&utm_medium=organic","carrier":"3","deviceFingerPrintId":"ffffffff-9333-3215-0000-00000625e8bb","date1":"2017-03-31_094538+0100","af_preinstalled":"false","advertiserIdEnabled":"false","iaecounter":"83","appsflyerKey":"ud****************WMsN","lang_code":"en","installer_package":"com.android.vending","app_version_name":"4.2.0","lang":"English","timepassedsincelastlaunch":"0","advertiserId":"c03e4a93-2cfe-4959-9e0f-f2cc22eb4a3f","isGaidWithGps":"true","deviceData":{"build_display_id":"N4F26T","arch":"","cpu_abi":"arm64-v8a","cpu_abi2":""},"af_v2":"0ef0ceec1b26c1024956dd1400668946a377ebd6","deviceType":"user","af_v":"6dcf340e247b0e5eefc85969eb6a1719ef847962","app_version_code":"22","af_events_api":"1","platformextension":"android_native","network":"WIFI","operator":"","country":"GB","date2":"2017-03-31_094538+0100","brand":"google","af_timestamp":"1491297894427","uid":"1490949938374-6358470765804798556","isFirstCall":"false","fb":"001217d2-982f-401d-9d3c-3ed8c9cc0eff","af_preinstall_name":"af","counter":"250","model":"Nexus 6P","product":"angler"}

# From second successful run (still in background)

04-04 10:25:11.306 22384-23099/? I/AppsFlyer_4.6.7: onBecameBackground
04-04 10:25:11.307 22384-23099/? I/AppsFlyer_4.6.7: callStatsBackground background call
04-04 10:25:11.307 22384-23099/? I/AppsFlyer_4.6.7: app went to background
04-04 10:25:11.314 22384-23099/? I/AppsFlyer_4.6.7: call = https://stats.appsflyer.com/stats; size = 430 bytes; body = {"advertiserId":"c03e4a93-2cfe-4959-9e0f-f2cc22eb4a3f","platform":"Android","gcd_conversion_data_timing":"0","devkey":"ud****************WMsN","statType":"user_closed_app","originalAppsflyerId":"1490949938374-6358470765804798556","time_in_app":"17","deviceFingerPrintId":"ffffffff-9333-3215-0000-00000625e8bb","app_id":"com.appsflyer.referrerSender","uid":"1490949938374-6358470765804798556","channel":null,"launch_counter":"250"}
04-04 10:25:11.493 22384-23099/? I/AppsFlyer_4.6.7: Status 200 ok
04-04 10:25:11.493 22384-23099/? I/AppsFlyer_4.6.7: Connection call succeeded
04-04 10:25:12.564 22384-22384/? I/AppsFlyer_4.6.7: onBecameForeground
04-04 10:25:12.715 22384-23352/? I/AppsFlyer_4.6.7: sendTrackingWithEvent from activity: com.appsflyer.referrerSender.ResultActivity
04-04 10:25:12.715 22384-23352/? I/AppsFlyer_4.6.7: Trying to fetch GAID..
04-04 10:25:12.727 22384-23352/? I/AppsFlyer_4.6.7: ******* sendTrackingWithEvent: Launch
04-04 10:25:12.756 22384-23352/? I/AppsFlyer_4.6.7: IMEI was not collected.
04-04 10:25:12.757 22384-23352/? I/AppsFlyer_4.6.7: Android ID was not collected.
04-04 10:25:12.764 22384-23352/? I/AppsFlyer_4.6.7: AppsFlyer: first launch date: 2017-03-31_095752+0100
04-04 10:25:12.764 22384-23352/? I/AppsFlyer_4.6.7: AppsFlyerLib.sendTrackingWithEvent
04-04 10:25:12.765 22384-23352/? I/AppsFlyer_4.6.7: url: https://t.appsflyer.com/api/v4/androidevent?buildnumber=4.6.7&app_id=com.appsflyer.referrerSender
04-04 10:25:12.766 22384-23352/? I/AppsFlyer_4.6.7: data: {"device":"angler","firstLaunchDate":"2017-03-31_095752+0100","installDate":"2017-03-31_094538+0100","sdk":"25","referrer":"utm_source=google-play&utm_medium=organic","carrier":"3","deviceFingerPrintId":"ffffffff-9333-3215-0000-00000625e8bb","date1":"2017-03-31_094538+0100","af_preinstalled":"false","advertiserIdEnabled":"false","iaecounter":"83","appsflyerKey":"ud****************WMsN","lang_code":"en","installer_package":"com.android.vending","app_version_name":"4.2.0","lang":"English","timepassedsincelastlaunch":"18","advertiserId":"c03e4a93-2cfe-4959-9e0f-f2cc22eb4a3f","isGaidWithGps":"true","deviceData":{"build_display_id":"N4F26T","arch":"","cpu_abi":"arm64-v8a","cpu_abi2":""},"af_v2":"1a2479e29b108935f65c71c53fa8171d7c2bb39b","deviceType":"user","af_v":"ffe03263b167762dd2679a8d898f3580010cf767","app_version_code":"22","af_events_api":"1","platformextension":"android_native","network":"WIFI","operator":"","country":"GB","date2":"2017-03-31_094538+0100","brand":"google","af_timestamp":"1491297912727","uid":"1490949938374-6358470765804798556","isFirstCall":"false","fb":"001217d2-982f-401d-9d3c-3ed8c9cc0eff","af_preinstall_name":"af","counter":"251","model":"Nexus 6P","product":"angler"}
```

**Comment by davidxmoody** [comment]
Another update: We decided to release to the Play Store anyway. We are still getting the same behaviour (tested on a whitelisted device):

- Clicks are always tracked correctly
- Installing the app and launching it once (and playing around in it) does not record an install
- Swiping away the app and re-launching it does not record an install
- Re-opening the app while it was already running in the background does record an install

This was tested by generating a tracking link, waiting a minute between doing each step and refreshing the AppsFlyer dashboard in between steps.

We can live with the behaviour for now (because if someone does not come back to the app then it isn't such a valuable install anyway) but it would be really nice to figure out what's going on.

**Comment by AbhishekCode** [comment]
need better documentation for testing. For a paid service like these proper documentation is required. While they respond quickly in support email, but those conversations are often useless and unproductive.  At this point I'm not even sure whether this sdk works with react native or not. Only link click is recorded which pretty simple thing anyway.

**Comment by af-fess** [comment]
@davidxmoody @AbhishekCode the problem exists in  integration testing app itself (timeout issue), not in SDK. it will be published soon (the fix already done).

**Comment by af-fess** [comment]
Last week we released `1.4.1` with the fix on Android side + separation of `getConversionData` and `onAppOpenAttribution` callbacks. This change should fix this issue.

**Comment by AbdulMateen-Tintash** [comment]
Hi, I'm facing the same issue on Android. Can you please confirm if it was fixed or not?
